### PR TITLE
6317 agility laps

### PR DESCRIPTION
**Pull request by mfedirko**
Fix issue #6317

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilityPlugin.java [hunk 1]
```diff
@@ -43,24 +43,7 @@
 import net.runelite.api.Tile;
 import net.runelite.api.TileObject;
 import net.runelite.api.coords.WorldPoint;
-import net.runelite.api.events.ConfigChanged;
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilityPlugin.java]: You need to fix your imports according to the wiki

https://github.com/runelite/runelite/wiki/Code-Conventions

File: runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilityPlugin.java [hunk 2]
```diff
@@ -187,23 +170,28 @@ public void onExperienceChanged(ExperienceChanged event)
...
+
+			session = new AgilitySession(course);
+			// New course found, reset lap count and set new course
+			session.resetLapCount();
 		}
 
-		if (session != null && session.getCourse() == course)
+		// Refresh laps till level even if did not complete course (check for mid course level up)
+		// Increment lap count if course completed
+		if (course.getCourseEndWorldPoints().length == 0
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilityPlugin.java]: Break this into logical if statements

File: runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilityPlugin.java [hunk 3]
```diff
@@ -187,23 +170,28 @@ public void onExperienceChanged(ExperienceChanged event)
...
-		if (course == null
-			|| (course.getCourseEndWorldPoints().length == 0
-			? Math.abs(course.getLastObstacleXp() - skillGained) > 1
-			: Arrays.stream(course.getCourseEndWorldPoints()).noneMatch(wp -> wp.equals(client.getLocalPlayer().getWorldLocation()))))
-		{
+		if (course == null)
 			return;
+
+		// Get new agility session
+		if ( !(session != null && session.getCourse() == course))
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilityPlugin.java]: This seems convoluted... Why not `session == null || session.getCourse() != course`?

File: runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilitySession.java [hunk 4]
```diff
@@ -49,7 +49,12 @@ void incrementLapCount(Client client)
 	{
 		lastLapCompleted = Instant.now();
 		++totalLaps;
+		refreshLapsTillLvl(client);
 
+	}
+
+	void refreshLapsTillLvl(Client client)
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilitySession.java]: `calculateLapsToLevel` or `reloadLapsToLevel` would be a better name.

**Comment by deathbeam** [comment]
Better will be probably 1. move the agi lap tracking code to the XP tracker as ActionType, 2. make agi plugin use it, 3. with that change the problem should solve itself

File: runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilityPlugin.java (hunk 3 above)
Comment (mfedirko) [reply runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilityPlugin.java]: resolved in latest commit

File: runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilitySession.java (hunk 4 above)
Comment (mfedirko) [reply runelite-client/src/main/java/net/runelite/client/plugins/agility/AgilitySession.java]: renamed in latest commit

**Pull request closed without merging**